pkg/controller/controller.go

```diff
@@ -0,0 +1,288 @@
+		remoteServiceMap: c.remoteServiceMap,
+		federator:        c.federator,
+		processingMutex:  c.processingMutex,
+	}
+	c.processingMutex.Lock()
+	defer c.processingMutex.Unlock()
+	c.remoteClusters[clusterID] = remoteCluster
+
+	serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+	})
+
+	go remoteCluster.run()
+}
+
+func (c *LightHouseController) removeServiceWatcher(clusterID string) {
+	remoteCluster, ok := c.remoteClusters[clusterID]
+	if ok {
+		klog.V(2).Infof("stopping watcher for %s", clusterID)
+		remoteCluster.close()
+		delete(c.remoteClusters, clusterID)
+	}
+}
+
+func (rc *remoteCluster) close() {
+	close(rc.stopCh)
+	rc.queue.ShutDown()
+	klog.Infof(" watcher queue shutdown for cluster: %s", rc.clusterID)
+}
+
+func (c *remoteCluster) run() {
+	defer utilruntime.HandleCrash()
+
+	go c.serviceInformer.Run(c.stopCh)
+
+	if !cache.WaitForCacheSync(c.stopCh, c.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return
+	}
+
+	go c.runWorker()
+}
+
+// HasSynced is required for the cache.Controller interface.
+func (c *remoteCluster) HasSynced() bool {
+	return c.serviceInformer.HasSynced()
+}
+
+func (c *remoteCluster) runWorker() {
+	for {
+		key, shutdown := c.queue.Get()
+		if shutdown {
+			klog.Infof("Lighthouse watcher for cluster \"%s\" stopped", c.clusterID)
+			return
+		}
+		defer c.queue.Done(key)
+
+		obj, exists, err := c.serviceInformer.GetIndexer().GetByKey(key.(string))
+
+		if err != nil {
+			klog.Errorf("Error processing %s (will retry): %v", key, err)
+			// requeue the item to work on later
+			c.queue.AddRateLimited(key)
+			continue
+		}
+		if !exists {
+			err := c.serviceDeleted(key.(string))
+			if err != nil {
+				klog.Errorf("Error deleting service in cluster: %v", err)
+				c.queue.AddRateLimited(key)
+			}
+		} else {
+			err := c.serviceCreated(obj, c.clusterID)
+			if err != nil {
+				klog.Errorf("Error creating service in cluster: %v", err)
+				c.queue.AddRateLimited(key)
+			}
+		}
+		c.queue.Forget(key)
+	}
+}
+
+func (c *remoteCluster) serviceCreated(obj interface{}, clusterID string) error {
+	klog.V(2).Infof("ServiceHandler.ObjectCreated")
+	service := obj.(*core_v1.Service)
+	key := service.ObjectMeta.Name + ":" + service.ObjectMeta.Namespace
+	if rs, exists := c.remoteServiceMap[key]; !exists {
+		c.processingMutex.Lock()
+		defer c.processingMutex.Unlock()
+		c.remoteServiceMap[key] = c.createLightHouseCRD(service, clusterID)
+		klog.Infof("A new service is created and it is added to the map: %s", key)
+	} else {
+		klog.Infof("An addService event was received for a Service already in our cache: %s, updating instead", key)
```

Comment:
How about `"Adding service from cluster %q to existing MultiClusterService %q", clusterID, key`?